File: day02/part2.swift
func isSafe(_ report: [Int]) -> Bool {
    var isIncreasing = true
    var isDecreasing = true
    var inRange = true

    for i in 1..<report.count {
        let delta = report[i] - report[i - 1]

        isIncreasing = isIncreasing && delta > 0
        isDecreasing = isDecreasing && delta < 0
        inRange = inRange && (1...3).contains(abs(delta))
    }

    return (isIncreasing || isDecreasing) && inRange
}

var safeCount = 0
while let line = readLine() {
    let report = line.split(separator: " ").map { Int($0)! }
    for i in 0..<report.count {
        var reduced = report
        reduced.remove(at: i)
        if isSafe(reduced) {
            safeCount += 1
            break
        }
    }
}

print(safeCount)

File: day02/part2.js
function isSafe(report) {
    let [previous, ...levels] = report

    let increasing = true;
    let decreasing = true;
    let minDelta = Number.MAX_SAFE_INTEGER;
    let maxDelta = 0;

    for (const level of levels) {
        const delta = level - previous;

        increasing = increasing && delta > 0;
        decreasing = decreasing && delta < 0;
        maxDelta = Math.max(maxDelta, Math.abs(delta));
        minDelta = Math.min(minDelta, Math.abs(delta));  
        
        previous = level;
    }

    return (increasing || decreasing) && minDelta >= 1 && maxDelta <= 3;
}

let safeCount = 0;
for await (const line of console) {
    const report = line.split(" ").map(Number);
    
    for (let i = -1; i < report.length; i += 1) {
        let permutation = report.filter((_, index) => index !== i);
        if (isSafe(permutation)) {
            safeCount += 1;
            break;
        }
    }
}

console.log(safeCount);